### PR TITLE
fix: improve text wrapping in ellipse and alignment

### PR DESCRIPTION
**Pull request by ad1992**
Partially fixes https://github.com/excalidraw/excalidraw/issues/5671

## Text Wrapping Inside Ellipse
I am taking the largest rectangle inside the ellipse whose dimensions are equal to 
```
width = Math.round((ellipse.width / 2) * Math.sqrt(2)) - BOUND_TEXT_PADDING * 2
height = Math.round((ellipse.height / 2) * Math.sqrt(2)) - BOUND_TEXT_PADDING * 2
```
For derivation you can refer [here](https://math.stackexchange.com/a/240201)
Screenshot 2023-02-17 at 1 36 44 PM

https://user-images.githubusercontent.com/11256141/215434156-fb2d3075-859e-4193-86ec-b82aeea8e9ee.mp4

Screenshot 2023-02-17 at 3 51 25 PM

## Calculating the coordinates 
Screenshot 2023-02-21 at 12 27 59 PM

## Alignment Fixes

Also, some alignment fixes along with this (added same padding as it is when bottom aligned)

### Vertical top align

image

### Always compute height in redrawBoundingBox

This fixes the case where height was getting updated due to text overflow when updating font properties but alignment was wrong for vertical top/bottom alignment as we we were only calculating height when center aligning - https://github.com/excalidraw/excalidraw/issues/6200

 

https://user-images.githubusercontent.com/11256141/219652817-f2ac7bf4-f4a3-4c93-945d-6870ad6f4528.mov

- [x] compute coords correctly when alignment changes
- [x] updated dims when font properties updated
- [x] Compute coords correctly when resize
- [x] Create a util to compute `boundTextElement` position and reuse
- [x] Fix tests
- [x] Add newtests

### CODE REVIEW

File: src/element/textElement.ts [hunk 1]
```diff
@@ -44,68 +44,76 @@ export const redrawTextBoundingBox = (
...
-  let text = textElement.text;
+
+  const boundTextUpdates: {
+    x: number;
+    y: number;
+    text: string;
+    width: number;
+    height: number;
+    baseline: number;
+  } = {
```
Comment (dwelle) [inline src/element/textElement.ts:55]: type annotation any useful here?

File: src/element/textElement.ts [hunk 2]
```diff
@@ -221,44 +233,51 @@ export const handleBindTextResize = (
       baseline: nextBaseLine,
     });
     if (!isArrowElement(container)) {
-      updateBoundTextPosition(
+      const { x, y } = computeBoundTextPosition(
         container,
         textElement as ExcalidrawTextElementWithContainer,
       );
+      mutateElement(textElement, { x, y });
```
Comment (dwelle) [inline src/element/textElement.ts:240]: I'd get in the habit of not destructuring + initializing variables + constructing new objects, if not necessary. Here, you can just pass the `computeBoundTextPosition()` as 2nd arg to `mtuateElement` directly.

Not critical, but still somewhat a hot path.

**Comment by dwelle** [comment]
One thing, if you paste an ellipse that was using the old wrapping algo, it requires 2x edits to wrap right:

excal-ellipse-wrapping-double-confirm

```
{"type":"excalidraw/clipboard","elements":[{"id":"NHdwMPZKgWJPtNn9DZGRH","type":"ellipse","x":-3857.193168618643,"y":-4795.797455080002,"width":234,"height":230.3864066699789,"angle":0,"strokeColor":"#5c940d","backgroundColor":"#fab00541","fillStyle":"solid","strokeWidth":2,"strokeStyle":"solid","roughness":1,"opacity":100,"groupIds":[],"roundness":null,"seed":826851654,"version":14,"versionNonce":145058630,"isDeleted":false,"boundElements":[{"type":"text","id":"WUz0xm9C4THfmw5U6q1vH"}],"updated":1676897139026,"link":null,"locked":false},{"id":"WUz0xm9C4THfmw5U6q1vH","type":"text","x":-3851.9999652836536,"y":-4770.604251745012,"width":220,"height":180,"angle":0,"strokeColor":"#5c940d","backgroundColor":"#fab00541","fillStyle":"solid","strokeWidth":2,"strokeStyle":"solid","roughness":1,"opacity":100,"groupIds":[],"roundness":null,"seed":1668500954,"version":218,"versionNonce":1284534362,"isDeleted":false,"boundElements":null,"updated":1676897143111,"link":null,"locked":false,"text":"asdkajsdk jasdkj askjdkaj \nsdkasjd kasdj kadjkajksdj \nkasjdk asjdkasj kasjd \njksajdk ajskdj kasdjkasjd \naksdjk ajdk ajskdjaskd \njaksdj aksdjaksjdkasjdka \nsjdkasjdk ajsdas dasdas \ndas das d asdas asd asd \nasdas dasdasd","fontSize":16,"fontFamily":1,"textAlign":"center","verticalAlign":"middle","baseline":174,"containerId":"NHdwMPZKgWJPtNn9DZGRH","originalText":"asdkajsdk jasdkj askjdkaj sdkasjd kasdj kadjkajksdj kasjdk asjdkasj kasjd jksajdk ajskdj kasdjkasjd aksdjk ajdk ajskdjaskd jaksdj aksdjaksjdkasjdka sjdkasjdk ajsdas dasdas das das d asdas asd asd asdas dasdasd"}],"files":{}}
```

**Comment by ad1992** [comment]
> One thing, if you paste an ellipse that was using the old wrapping algo, it requires 2x edits to wrap right:
> 
> excal-ellipse-wrapping-double-confirm [ excal-ellipse-wrapping-double-confirm ](https://user-images.githubusercontent.com/5153846/220112499-c5a3dbc4-1df0-43d5-ac99-20e0c6e804dd.gif) [ ](https://user-images.githubusercontent.com/5153846/220112499-c5a3dbc4-1df0-43d5-ac99-20e0c6e804dd.gif)
> 
> ```
> {"type":"excalidraw/clipboard","elements":[{"id":"NHdwMPZKgWJPtNn9DZGRH","type":"ellipse","x":-3857.193168618643,"y":-4795.797455080002,"width":234,"height":230.3864066699789,"angle":0,"strokeColor":"#5c940d","backgroundColor":"#fab00541","fillStyle":"solid","strokeWidth":2,"strokeStyle":"solid","roughness":1,"opacity":100,"groupIds":[],"roundness":null,"seed":826851654,"version":14,"versionNonce":145058630,"isDeleted":false,"boundElements":[{"type":"text","id":"WUz0xm9C4THfmw5U6q1vH"}],"updated":1676897139026,"link":null,"locked":false},{"id":"WUz0xm9C4THfmw5U6q1vH","type":"text","x":-3851.9999652836536,"y":-4770.604251745012,"width":220,"height":180,"angle":0,"strokeColor":"#5c940d","backgroundColor":"#fab00541","fillStyle":"solid","strokeWidth":2,"strokeStyle":"solid","roughness":1,"opacity":100,"groupIds":[],"roundness":null,"seed":1668500954,"version":218,"versionNonce":1284534362,"isDeleted":false,"boundElements":null,"updated":1676897143111,"link":null,"locked":false,"text":"asdkajsdk jasdkj askjdkaj \nsdkasjd kasdj kadjkajksdj \nkasjdk asjdkasj kasjd \njksajdk ajskdj kasdjkasjd \naksdjk ajdk ajskdjaskd \njaksdj aksdjaksjdkasjdka \nsjdkasjdk ajsdas dasdas \ndas das d asdas asd asd \nasdas dasdasd","fontSize":16,"fontFamily":1,"textAlign":"center","verticalAlign":"middle","baseline":174,"containerId":"NHdwMPZKgWJPtNn9DZGRH","originalText":"asdkajsdk jasdkj askjdkaj sdkasjd kasdj kadjkajksdj kasjdk asjdkasj kasjd jksajdk ajskdj kasdjkasjd aksdjk ajdk ajskdjaskd jaksdj aksdjaksjdkasjdka sjdkasjdk ajsdas dasdas das das d asdas asd asd asdas dasdasd"}],"files":{}}
> ```

Should be fine now when refreshing and pasting from older version.

File: src/element/newElement.ts [hunk 3]
```diff
@@ -290,6 +290,8 @@ export const getMaxContainerWidth = (container: ExcalidrawElement) => {
       return BOUND_TEXT_PADDING * 8 * 2;
     }
     return containerWidth;
+  } else if (container.type === "ellipse") {
+    return Math.round((width / 2) * Math.sqrt(2)) - BOUND_TEXT_PADDING * 2;
```
Comment (dwelle) [inline src/element/newElement.ts:294]: can you add comment on what's the significance of multiplying by `1.41`? Seems a bit arbitary.

File: src/element/textElement.ts [hunk 4]
```diff
@@ -621,6 +635,23 @@ export const getContainerCenter = (
   return { x: midSegmentMidpoint[0], y: midSegmentMidpoint[1] };
 };
 
+export const getContainerCoords = (container: NonDeletedExcalidrawElement) => {
+  if (container.type === "ellipse") {
+    const offsetX =
+      (container.width / 2) * (1 - Math.sqrt(2) / 2) + BOUND_TEXT_PADDING;
```
[the remark this answers is not in the document]
Comment (ad1992) [reply src/element/textElement.ts:642]: I have added the explanation in PR desc and added PR link in the comment

**Pull request merged**